File: web/src/api/history.ts
import { apiFetch } from './client';
import type { RunRecord, HistoryStats, TrendPoint, RunComparison } from '../types';

export const historyApi = {
  list: (opts?: {
    query?:       string;
    table?:       string;
    type?:        string;
    limit?:       number;
    offset?:      number;
    top?:         boolean;
    regressions?: boolean;
  }): Promise<RunRecord[]> => {
    const params = new URLSearchParams();
    if (opts?.query)       params.set('query',       opts.query);
    if (opts?.table)       params.set('table',       opts.table);
    if (opts?.type)        params.set('type',        opts.type);
    if (opts?.limit  != null) params.set('limit',   String(opts.limit));
    if (opts?.offset != null) params.set('offset',  String(opts.offset));
    if (opts?.top)         params.set('top',         'true');
    if (opts?.regressions) params.set('regressions', 'true');
    const qs = params.toString();
    return apiFetch<RunRecord[]>(`/api/history${qs ? `?${qs}` : ''}`);
  },

  stats: (): Promise<HistoryStats> =>
    apiFetch<HistoryStats>('/api/history/stats'),

  trend: (opts: { table?: string; query_hash?: string }): Promise<TrendPoint[]> => {
    const params = new URLSearchParams();
    if (opts.table)       params.set('table',       opts.table);
    if (opts.query_hash)  params.set('query_hash',  opts.query_hash);
    return apiFetch<TrendPoint[]>(`/api/history/trend?${params.toString()}`);
  },

  compare: (a: number, b: number): Promise<RunComparison> =>
    apiFetch<RunComparison>(`/api/history/compare?a=${a}&b=${b}`),

  get: (id: number): Promise<RunRecord> =>
    apiFetch<RunRecord>(`/api/history/${id}`),

  delete: (id: number): Promise<{ deleted: boolean; id: number }> =>
    apiFetch<{ deleted: boolean; id: number }>(`/api/history/${id}`, {
      method: 'DELETE',
    }),
};

File: bridge/routers/history.py
from fastapi import APIRouter, Query, HTTPException
from tools.history import (
    get_history,
    get_stats,
    get_trend,
    get_regressions,
    get_top_improvements,
    compare_runs,
)

router = APIRouter(prefix="/api/history", tags=["history"])


@router.get("")
def list_history(
    query:       str  = Query(None),
    table:       str  = Query(None),
    type:        str  = Query(None),
    limit:       int  = Query(50),
    offset:      int  = Query(0),
    top:         bool = Query(False),
    regressions: bool = Query(False),
):
    if top:
        return get_top_improvements(limit=limit)
    if regressions:
        return get_regressions()

    # Fetch with enough headroom to support offset slicing
    runs = get_history(query=query, table_name=table, limit=limit + offset)

    # Filter by run_type if provided
    if type:
        runs = [r for r in runs if r.get("run_type") == type]

    return runs[offset:]


@router.get("/stats")
def history_stats():
    return get_stats()


@router.get("/trend")
def trend(
    table:      str = Query(None),
    query_hash: str = Query(None),
):
    if not table and not query_hash:
        raise HTTPException(400, "Provide table or query_hash")
    return get_trend(table_name=table, query_hash=query_hash)


@router.get("/compare")
def compare(
    a: int = Query(...),
    b: int = Query(...),
):
    result = compare_runs(a, b)
    if "error" in result:
        raise HTTPException(404, result["error"])
    return result


@router.get("/{run_id}")
def get_run(run_id: int):
    # Pull a broad slice and find the matching ID
    runs = get_history(limit=5000)
    match = next((r for r in runs if r["id"] == run_id), None)
    if not match:
        raise HTTPException(404, f"Run {run_id} not found")
    return match


@router.delete("/{run_id}")
def delete_run(run_id: int):
    from tools.history import _get_conn
    conn = _get_conn()
    conn.execute("DELETE FROM runs WHERE id = ?", (run_id,))
    conn.commit()
    conn.close()
    return {"deleted": True, "id": run_id}
